plugins/lang-plan-a/src/main/java/org/elasticsearch/plan/a/Metadata.java

```diff
@@ -0,0 +1,598 @@
+
+        /**
+         * The target variable is set to a value based on the type of ANTLR node that is visited.  This is used by
+         * {@link Writer} to determine whether a cast, store, load, or method call should be written in byte code
+         * depending on what the target variable is.
+         */
+        Object target = null;
+
+        /**
+         * The last variable is set to true when the last ANTLR node of the variable/method chain is visted.  This is
+         * used by the {@link Writer} in conjuction with the storeExpr variable to determine whether or not a store
+         * needs to be written as opposed to a load.
+         */
+        boolean last = false;
+
+        /**
+         * The type variable is set to the type that a visited node ends with.  This is used by both the
+         * {@link Analyzer} and {@link Writer} to make decisions about compound assignments, String constants, and
+         * shortcuts.
+         */
+        Type type = null;
+
+        /**
+         * The promote variable is set to the type of a promotion within a compound assignment.  Compound assignments
+         * may require promotion between the left-hand side variable and right-hand side value.  This is used by the
+         * {@link Writer} to make the correct decision about the byte code operation.
+         */
+        Type promote = null;
+
+        /**
+         * The castFrom variable is set during a compound assignment.  This is used by the {@link Writer} to
+         * cast the values to the promoted type during a compound assignment.
+         */
+        Cast castFrom = null;
+
+        /**
+         * The castTo variable is set during an explicit cast in a variable/method chain or during a compound
+         * assignment.  This is used by the {@link Writer} to either do an explicit cast, or cast the values
+         * from the promoted type back to the original type during a compound assignment.
+         */
+        Cast castTo = null;
+
+        // The constructor takes in the top-level ANTLR node for variable/method chain and the associated ANTLR node.
+        private ExtNodeMetadata(final ParserRuleContext parent, final ParserRuleContext source) {
+            this.parent = parent;
+            this.source = source;
+        }
+    }
+
+    /**
+     * A utility method to output consistent error messages.
+     * @param ctx The ANTLR node the error occurred in.
+     * @return The error message with tacked on line number and character position.
+     */
+    static String error(final ParserRuleContext ctx) {
+        return "Error [" + ctx.getStart().getLine() + ":" + ctx.getStart().getCharPositionInLine() + "]: ";
+    }
+
+    /**
+     * The definition variable acts as both the Plan A API and white-list for what types and methods are allowed.
+     */
+    final Definition definition;
+
+    /**
+     * The source variable is the original text of the input script.  This is used to write out the source code into
```

Comment:
I'd remove "The X variable is" and keep the rest of the comment. Sometimes *what* the variable is is obvious from the type and name but its super helpful for you to explain why you keep it around like you do in this case.